examples/double_gyre.jl

```diff
@@ -0,0 +1,292 @@
+# # Double Gyre
+#
+# This example simulates a double gyre following:
+# https://mitgcm.readthedocs.io/en/latest/examples/baroclinic_gyre/baroclinic_gyre.html
+
+using Oceananigans
+using Oceananigans.Units
+using Oceananigans.Grids: xnode, ynode, znode
+using Oceananigans.TurbulenceClosures: VerticallyImplicitTimeDiscretization
+using Plots
+using Printf
+
+architecture = GPU()
+
+const Lx = 4000kilometers # east-west extent [m]
+const Ly = 6000kilometers # north-south extent [m]
+const Lz = 1.8kilometers # depth [m]
+
+Δt₀ = 10minutes
+stop_time = 10years
+
+Nx = 160
+Ny = 240
+Nz = 50
+
+σ = 1.2 # stretching factor
+hyperbolically_spaced_faces(k) = - Lz * (1 - tanh(σ * (k - 1) / Nz) / tanh(σ))
+
+grid = RectilinearGrid(architecture;
+                           size = (Nx, Ny, Nz),
+                           halo = (3, 3, 3),
+                              x = (-Lx/2, Lx/2),
+                              y = (-Ly/2, Ly/2),
+                              z = hyperbolically_spaced_faces,
+                       topology = (Bounded, Bounded, Bounded))
+
+# plot(grid.Δzᵃᵃᶜ[1:Nz], grid.zᵃᵃᶜ[1:Nz],
+#       marker = :circle,
+#       ylabel = "Depth (m)",
+#       xlabel = "Vertical spacing (m)",
+#       legend = nothing)
+
+α  = 2e-4 # [K⁻¹] thermal expansion coefficient
+g  = 9.81 # [m s⁻²] gravitational constant
+cᵖ = 3991 # [J K⁻¹ kg⁻¹] heat capacity for seawater
+ρ₀ = 1028 # [kg m⁻³] reference density
+
+parameters = (Ly = Ly,
+              Lz = Lz,
+               τ = 0.1 / ρ₀,     # surface kinematic wind stress [m² s⁻²]
+               μ = 1 / 30days,   # bottom drag damping time-scale [s⁻¹]
+              Δb = 30 * α * g,   # surface vertical buoyancy gradient [s⁻²]
+               λ = 30days        # relaxation time scale [s]
+              )
+
+# ## Boundary conditions
+#
+# ### Wind stress
+
+@inline function u_stress(i, j, grid, clock, model_fields, p)
+    y = ynode(Center(), j, grid)
+    return - p.τ * cos(2π * y / p.Ly)
+end
+
+u_stress_bc = FluxBoundaryCondition(u_stress, discrete_form=true, parameters=parameters)
+
+# ### Bottom drag
+
+@inline u_drag(i, j, grid, clock, model_fields, p) = @inbounds - p.μ * p.Lz * model_fields.u[i, j, 1]
+@inline v_drag(i, j, grid, clock, model_fields, p) = @inbounds - p.μ * p.Lz * model_fields.v[i, j, 1]
+
+u_drag_bc = FluxBoundaryCondition(u_drag, discrete_form=true, parameters=parameters)
+v_drag_bc = FluxBoundaryCondition(v_drag, discrete_form=true, parameters=parameters)
+
+u_bcs = FieldBoundaryConditions(top = u_stress_bc, bottom = u_drag_bc)
+v_bcs = FieldBoundaryConditions(bottom = v_drag_bc)
+
+
+# ### Buoyancy relaxation
+
+@inline function buoyancy_relaxation(i, j, k, grid, clock, model_fields, p)
+   y = ynode(Center(), j, grid)
+   b = @inbounds model_fields.b[i, j, k]
+   return - 1 / p.λ * (b - p.Δb * y / p.Ly)
+end
+
+Fb = Forcing(buoyancy_relaxation, discrete_form = true, parameters = parameters)
```

Comment:
Seems like these would be cleaner if we used the continuous form?

Reply:
Agreed!

Reply:
Done in [00c2a69](https://github.com/CliMA/Oceananigans.jl/pull/1085/commits/00c2a698016e2f0dffe87b22e5ece64b23689648)